crates/rust-analyzer/src/lsp_ext.rs

```diff
@@ -376,11 +376,28 @@ pub struct SnippetTextEdit {
 pub enum HoverRequest {}
 
 impl Request for HoverRequest {
-    type Params = lsp_types::HoverParams;
+    type Params = HoverParams;
     type Result = Option<Hover>;
     const METHOD: &'static str = "textDocument/hover";
 }
 
+#[derive(Debug, Eq, PartialEq, Clone, Deserialize, Serialize)]
+#[serde(rename_all = "camelCase")]
+pub struct HoverParams {
+    pub text_document: TextDocumentIdentifier,
+    pub position: PositionOrRange,
+
```

Comment:
Hm, I am having second thoughts here. Perhaps 

```
pub position: Option<Position>,
pub range: Option<Range>,
```

would be better from the protocol point of view? 

for us, this doesn't really matter, but I wonder if this extension might become used by more then one server. 


Might be a good idea to survery other LSP requests and see what would be the best JSON shape to fit with the rest of the protocol.

Reply:
Not sure if I missed that but wouldn't just a `Range` make more sense where the empty range = position?

Reply:
Clients still need to be able to pass a `Position` if they don't support the extension.

Reply:
Ah right we need this to be an extension to the original, conveniently forgot that.

Two fields for this feel odd to me, since having both set doesn't make that much sense while its still legal to do? Not sure what the advantages would be of having two fields over one that can be either a range or position.

Reply:
to clarify, what specifically feels odd to me is that the field called `position` can actually hold the `Range`.

Reply:
> to clarify, what specifically feels odd to me is that the field called `position` can actually hold the `Range`. 

That was the first thing that I was blocked by. I think that having separate props and fail if the request is invalid is a bit more clear for someone seeing this for the first time

Reply:
Thinking more about this, I think the current solution is fine.